### ISSUE TITLE
False positive: `Protocol member cannot be final` in non-instance assignment in a method

### ISSUE THREAD

**Issue by bzoracler** [python/mypy#17281]
**Bug Report, reproduction, & actual behaviour**

See [mypy Playground](https://mypy-play.net/?mypy=latest&python=3.12&flags=strict&gist=653c1686ec0f7b04a9b8c054c31fe6f3):

```python
import typing as t

class A(t.Protocol):
    def method(self) -> None:
        var: t.Final = 0  # mypy: Protocol member cannot be final  [misc]
```

**Expected Behavior**

No issues

**Your Environment**

- Mypy version used: 1.10
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.10, 3.12

**Comment by GiorgosPapoutsakis** [comment]
From my understanding, this behavior is a feature and not a bug. It is hard-coded and tested to ensure that protocol members cannot be marked as Final. Therefore, the expected behavior is correctly reported by mypy. Although this functionality can be restrictive in some use cases and maybe is worth considering a change. Should I try to make this modification?

**Comment by bzoracler** [comment]
This is a bug because `var` here is just a variable in the local function scope, not a protocol member. A protocol member would look like `self.var`.

**Comment by GiorgosPapoutsakis** [comment]
Thank you for the clarification
